src/models/tree/build_tree.h

```diff
@@ -102,9 +103,50 @@ inline __host__ __device__ auto ComputeHistogramBin(int node_id,
   return histogram_node == node_id;
 }
 
-__host__ __device__ inline auto CalculateLeafValue(double G, double H, double alpha) -> double
+namespace detail {
+__host__ __device__ inline auto CalculateLeafGain(const GPair& sum,
+                                                  double alpha_reg,
+                                                  double l1_regularization) -> double
 {
-  return -G / (H + alpha);
+  // l1 threshold
+  double z = 0.0;
+  if (sum.grad > l1_regularization) {
+    z = sum.grad - l1_regularization;
+  } else if (sum.grad < -l1_regularization) {
+    z = sum.grad + l1_regularization;
+  }
+  // Note: we could use the cuda __ddiv_rn intrinsic here if this becomes a bottleneck
+  return (z * z) / (sum.hess + alpha_reg);
```

Comment:
We could be tricky and use that `z * z` is always positive, but likely the math optimizer is able to even see that.

Reply:
I don't follow. Is there a way to do faster division if you know the numerator is positive?

Reply:
sorry, I was just thinking wrong, the comment was nonsense.